numpy/core/src/multiarray/item_selection.c

```diff
@@ -2364,6 +2377,9 @@ PyArray_Nonzero(PyArrayObject *self)
             Py_DECREF(ret);
             return NULL;
         }
+        
+        /* See if we need to check the error state due to iternext() */
+        needs_api |= NpyIter_IterationNeedsAPI(iter);
```

Comment:
I don't think this comment is correct - as far as I know `iternext` cannot fail.

Also, note that in `CountNonZero` we use `=` not `|=` (see #9300 for the diff). It would be good to be consistent, and my feeling is `=` is sufficient here.

Reply:
@garyg1 ping

Reply:
Yeah that makes sense, the IterationNeedsAPI includes the flag check for the data type anyway.

Reply:
Thanks @garyg1 this seems all right. If you have some time, it could be nice to add some test for these paths (because I could not find any quickly). Although, I think I would not bother for threading too much right now, just that we have code coverage at all.

Reply:
I added a test for proper handling of the error state. Seems like testing for GIL release would be a good idea, but the code is so simple that it's just an issue of whether the macros are correct.